edgelet/build/windows/build.ps1

```diff
@@ -1,23 +1,80 @@
 # Copyright (c) Microsoft. All rights reserved.
 
 param(
-    [switch]$Release
+    [switch]$Release,
+    [switch]$Arm
 )
 
 # Bring in util functions
 $util = Join-Path -Path $PSScriptRoot -ChildPath "util.ps1"
 . $util
 
 # Ensure rust is installed
-Assert-Rust
+Assert-Rust -Arm:$Arm
+
+$cargo = Get-CargoCommand -Arm:$Arm
+
+$ErrorActionPreference = 'Continue'
+
+# arm build has to use a few private forks of dependencies instead of the public ones, in order to to this, we have to 
+# 1. append a [patch] section in cargo.toml to use crate forks
+# 2. run cargo update commands to force update cargo.lock to use the forked crates
+# after these two steps, running cargo build for arm should not show any warning such as "Patch 'crate-x' was not used in the crate graph"
+if($Arm)
+{
+    $edgefolder = Get-EdgeletFolder
+    Set-Location -Path $edgefolder
+
+    $ForkedCrates = @"
+
+[patch.crates-io]
+backtrace = { git = "https://github.com/philipktlin/backtrace-rs", branch = "arm" }
+cmake = { git = "https://github.com/philipktlin/cmake-rs", branch = "arm" }
+dtoa = { git = "https://github.com/philipktlin/dtoa", branch = "arm" }
+iovec = { git = "https://github.com/philipktlin/iovec", branch = "arm" }
+mio = { git = "https://github.com/philipktlin/mio", branch = "arm" }
+miow = { git = "https://github.com/philipktlin/miow", branch = "arm" }
+serde-hjson = { git = "https://github.com/philipktlin/hjson-rust", branch = "arm" }
+winapi = { git = "https://github.com/philipktlin/winapi-rs", branch = "arm/v0.3.5" }
+
+[patch."https://github.com/Azure/mio-uds-windows"]
+mio-uds-windows = { git = "https://github.com/philipktlin/mio-uds-windows", branch = "arm" }
+
+[patch."https://github.com/Azure/hyperlocal-windows"]
+hyperlocal-windows = { git = "https://github.com/philipktlin/hyperlocal-windows", branch = "arm" }
+
+[patch."https://github.com/Azure/tokio-uds-windows"]
+tokio-uds-windows = { git = "https://github.com/philipktlin/tokio-uds-windows", branch = "arm" }
+"@
+
+    Write-Host "Append cargo.toml with $ForkedCrates"
+    Add-Content -Path cargo.toml -Value $ForkedCrates
+
+    Write-Host "Running cargo update to lock the crate forks required by arm build"
+    Invoke-Expression "$cargo update -p https://github.com/Azure/mio-uds-windows.git#mio-uds-windows:0.1.0"

```

Comment:
I assume you're downgrading mio because it happens to be upgraded to v0.6.16 by the previous `cargo update`, but there is no reason to assume that mio is the only one that needs to be downgraded. If mio got updated to v0.6.17 and that version also pulled in an updated version of some other dependency, then that second dependency would remain at the newer version.

Instead you should only need:

```powershell
Invoke-Expression "$cargo update -p winapi:0.3.5 --precise 0.3.5"
Invoke-Expression "$cargo update -p mio-uds-windows -p tokio-uds-windows"
```

Reply:
these two lines aren't enough, the patch list and the four update commands are the minimal I can get.
From my experience, with or without .git when referencing a crate does have impact to cargo update, it'll end up behaving differently.
Also, I need to cargo update to the specific version so the patch can take effect, because Philip's fork for mio is 0.6.14.

Reply:
>these two lines aren't enough, the patch list and the four update commands are the minimal I can get.

They're enough. I wrote my comment after I tested it on my machine.

>Also, I need to cargo update to the specific version so the patch can take effect, because Philip's fork for mio is 0.6.14.

The version in our lockfile is also 0.6.14.

Reply:
Did you cargo build after you updated the lock files? and on what branch? If I only patch mio-uds-windows, without patching tokio-uds-windows or hyperlocal-windows, after I ran your update commands then tried to build, the public mio-uds-windows was added back, which brought in winapi 0.2.8. the public mio-uds-windows was pulled in by hyperlocal-windows. I thought we're trying to get rid of 0.2.8 winapi...

```
name = "hyperlocal-windows"
version = "0.1.0"
source = "git+https://github.com/Azure/hyperlocal-windows#2bd432bbbfb5b1cf38429733dd9a593c7b97a850"
dependencies = [
 "futures 0.1.24 (registry+https://github.com/rust-lang/crates.io-index)",
 "hex 0.3.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "hyper 0.12.17 (registry+https://github.com/rust-lang/crates.io-index)",
 **"mio-uds-windows 0.1.0 (git+https://github.com/Azure/mio-uds-windows.git)",**

[[package]]
name = "mio-uds-windows"
version = "0.1.0"
source = "git+https://github.com/Azure/mio-uds-windows.git#87a4a9970e668fdbe9205f4039967e175182c70e"
dependencies = [
 "iovec 0.1.2 (git+https://github.com/philipktlin/iovec?branch=arm)",
 "kernel32-sys 0.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "lazycell 1.2.0 (registry+https://github.com/rust-lang/crates.io-index)",
 "log 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "mio 0.6.14 (git+https://github.com/philipktlin/mio?branch=arm)",
 "miow 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)",
 "winapi 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)",
 "ws2_32-sys 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)",
]
```

and here's the cargo build output, see first line

```
I:\iotedge-1\edgelet>cargo build --all --target thumbv7a-pc-windows-msvc --release
    Updating git repository `https://github.com/Azure/mio-uds-windows.git`
    Updating crates.io index
   Compiling winapi v0.2.8
   Compiling cmake v0.1.30 (https://github.com/philipktlin/cmake-rs?branch=arm#886c0dd7)
   Compiling regex-syntax v0.5.6
   Compiling openssl-sys v0.9.36
   Compiling winapi v0.3.5 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
   Compiling serde v1.0.84
   Compiling qu
```
and eventually build failed,

```
error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:86:19
   |
86 | pub type SIZE_T = ULONG_PTR;
   |                   ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:87:25
   |
87 | pub type PSIZE_T = *mut ULONG_PTR;
   |                         ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `LONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:88:20
   |
88 | pub type SSIZE_T = LONG_PTR;
   |                    ^^^^^^^^ not found in this scope

error[E0412]: cannot find type `LONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:89:26
   |
89 | pub type PSSIZE_T = *mut LONG_PTR;
   |                          ^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:90:22
   |
90 | pub type DWORD_PTR = ULONG_PTR;
   |                      ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:91:28
   |
91 | pub type PDWORD_PTR = *mut ULONG_PTR;
   |                            ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:98:22
   |
98 | pub type KAFFINITY = ULONG_PTR;
   |                      ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in the crate root
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\cfgmgr32.rs:43:28
   |
43 | pub type CONFLICT_LIST = ::ULONG_PTR;
   |                            ^^^^^^^^^ not found in the crate root

error[E0412]: cannot find type `TBBUTTON` in this scope
   --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\commctrl.rs:516:27
    |
516 | pub type PTBBUTTON = *mut TBBUTTON;
    |                           ^^^^^^^^ help: a type alias with a similar name exists: `PTBBUTTON`

error[E0412]: cannot find type `TBBUTTON` in this scope
   --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\commctrl.rs:517:28
    |
517 | pub type LPTBBUTTON = *mut TBBUTTON;
    |                            ^^^^^^^^ help: a type alias with a similar name exists: `PTBBUTTON`

error[E0412]: cannot find type `TBBUTTON` in this scope
   --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\commctrl.rs:518:31
    |
518 | pub type LPCTBBUTTON = *const TBBUTTON;
    |                               ^^^^^^^^ help: a type alias with a similar name exists: `PTBBUTTON`

error[E0412]: cannot find type `UINT_PTR` in the crate root
   --> C:\Users\dengy\.cargo\registry\s
```

Reply:
```powershell
$ git diff --patch HEAD~1

diff --git a/edgelet/Cargo.toml b/edgelet/Cargo.toml
index f86c77d6..33a16bb8 100644
--- a/edgelet/Cargo.toml
+++ b/edgelet/Cargo.toml
@@ -29,3 +29,16 @@ members = [
     "tokio-named-pipe",
     "win-logger",
 ]
+
+[patch.crates-io]
+backtrace = { git = "https://github.com/philipktlin/backtrace-rs", branch = "arm" }
+cmake = { git = "https://github.com/philipktlin/cmake-rs", branch = "arm" }
+dtoa = { git = "https://github.com/philipktlin/dtoa", branch = "arm" }
+iovec = { git = "https://github.com/philipktlin/iovec", branch = "arm" }
+mio = { git = "https://github.com/philipktlin/mio", branch = "arm" }
+miow = { git = "https://github.com/philipktlin/miow", branch = "arm" }
+serde-hjson = { git = "https://github.com/philipktlin/hjson-rust", branch = "arm" }
+winapi = { git = "https://github.com/philipktlin/winapi-rs", branch = "arm/v0.3.5" }
+
+[patch."https://github.com/Azure/mio-uds-windows.git"]
+mio-uds-windows = { git = "https://github.com/philipktlin/mio-uds-windows.git", branch = "arm" }
diff --git a/edgelet/build/windows/build.ps1 b/edgelet/build/windows/build.ps1
index 32628cad..d921e565 100644
--- a/edgelet/build/windows/build.ps1
+++ b/edgelet/build/windows/build.ps1
@@ -37,24 +37,16 @@ miow = { git = "https://github.com/philipktlin/miow", branch = "arm" }
 serde-hjson = { git = "https://github.com/philipktlin/hjson-rust", branch = "arm" }
 winapi = { git = "https://github.com/philipktlin/winapi-rs", branch = "arm/v0.3.5" }

-[patch."https://github.com/Azure/mio-uds-windows"]
+[patch."https://github.com/Azure/mio-uds-windows.git"]
 mio-uds-windows = { git = "https://github.com/philipktlin/mio-uds-windows", branch = "arm" }
-
-[patch."https://github.com/Azure/hyperlocal-windows"]
-hyperlocal-windows = { git = "https://github.com/philipktlin/hyperlocal-windows", branch = "arm" }
-
-[patch."https://github.com/Azure/tokio-uds-windows"]
-tokio-uds-windows = { git = "https://github.com/philipktlin/tokio-uds-windows", branch = "arm" }
 "@

     Write-Host "Append cargo.toml with $ForkedCrates"
     Add-Content -Path cargo.toml -Value $ForkedCrates

     Write-Host "Running cargo update to lock the crate forks required by arm build"
-    Invoke-Expression "$cargo update -p https://github.com/Azure/mio-uds-windows.git#mio-uds-windows:0.1.0"
-    Invoke-Expression "$cargo update -p mio --precise 0.6.14"
-    Invoke-Expression "$cargo update -p https://github.com/Azure/tokio-uds-windows.git#tokio-uds-windows:0.1.0"
-    Invoke-Expression "$cargo update -p winapi --precise 0.3.5"
+    Invoke-Expression "$cargo update -p winapi:0.3.5 --precise 0.3.5"
+    Invoke-Expression "$cargo update -p mio-uds-windows"
 }


diff --git a/edgelet/edgelet-http/Cargo.toml b/edgelet/edgelet-http/Cargo.toml
index 312c5239..ff155897 100644
--- a/edgelet/edgelet-http/Cargo.toml
+++ b/edgelet/edgelet-http/Cargo.toml
@@ -40,9 +40,9 @@ tokio-uds = "0.2"
 [target.'cfg(windows)'.dependencies]
 hyper-named-pipe = { path = "../hyper-named-pipe" }
 hyperlocal-windows = { git = "https://github.com/Azure/hyperlocal-windows" }
-mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows" }
+mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows.git" }
 tokio-named-pipe = { path = "../tokio-named-pipe" }
-tokio-uds-windows = { git = "https://github.com/Azure/tokio-uds-windows" }
+tokio-uds-windows = { git = "https://github.com/Azure/tokio-uds-windows.git" }
 winapi = { version = "0.3.5", features = ["winsock2"] }

 [dev-dependencies]
diff --git a/edgelet/edgelet-test-utils/Cargo.toml b/edgelet/edgelet-test-utils/Cargo.toml
index 82946a6c..fc6aa003 100644
--- a/edgelet/edgelet-test-utils/Cargo.toml
+++ b/edgelet/edgelet-test-utils/Cargo.toml
@@ -24,5 +24,5 @@ hyperlocal = "0.6"
 hyperlocal-windows = { git = "https://github.com/Azure/hyperlocal-windows" }
 mio = "0.6"
 mio-named-pipes = "0.1"
-mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows" }
+mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows.git" }
 miow = "0.3"



$ cargo update -p winapi:0.3.5 --precise 0.3.5

    Updating crates.io index
    Updating git repository `https://github.com/philipktlin/backtrace-rs`
    Updating git repository `https://github.com/philipktlin/cmake-rs`
    Updating git repository `https://github.com/philipktlin/dtoa`
    Updating git repository `https://github.com/philipktlin/iovec`
    Updating git repository `https://github.com/philipktlin/mio`
    Updating git repository `https://github.com/philipktlin/miow`
    Updating git repository `https://github.com/philipktlin/hjson-rust`
    Updating git repository `https://github.com/philipktlin/winapi-rs`
    Updating git repository `https://github.com/philipktlin/mio-uds-windows.git`
    Removing aho-corasick v0.5.3
      Adding backtrace v0.3.6 (https://github.com/philipktlin/backtrace-rs?branch=arm#a6caa2dd)
    Removing backtrace v0.3.6
      Adding backtrace-sys v0.1.16 (https://github.com/philipktlin/backtrace-rs?branch=arm#a6caa2dd)
    Removing backtrace-sys v0.1.16
      Adding cmake v0.1.30 (https://github.com/philipktlin/cmake-rs?branch=arm#886c0dd7)
    Removing cmake v0.1.30
      Adding dtoa v0.4.2 (https://github.com/philipktlin/dtoa?branch=arm#db97225f)
    Removing dtoa v0.4.2
      Adding iovec v0.1.2 (https://github.com/philipktlin/iovec?branch=arm#5d56185b)
    Removing iovec v0.1.2
    Removing kernel32-sys v0.2.2
    Removing memchr v0.1.11
      Adding mio v0.6.14 (https://github.com/philipktlin/mio?branch=arm#bd83233b)
    Removing mio v0.6.14
      Adding mio-uds-windows v0.1.0 (https://github.com/philipktlin/mio-uds-windows.git?branch=arm#581fcd3d)
    Removing mio-uds-windows v0.1.0 (https://github.com/Azure/mio-uds-windows.git#87a4a997)
      Adding miow v0.3.1 (https://github.com/philipktlin/miow?branch=arm#8f63c73f)
    Removing miow v0.2.1
    Removing miow v0.3.1
    Removing regex v0.1.80
    Removing regex-syntax v0.3.9
      Adding serde-hjson v0.8.1 (https://github.com/philipktlin/hjson-rust?branch=arm#e1fc93df)
    Removing serde-hjson v0.8.1
    Removing thread-id v2.0.0
    Removing thread_local v0.2.7
    Removing utf8-ranges v0.1.3
      Adding winapi v0.3.5 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
    Removing winapi v0.2.8
    Removing winapi v0.3.5
    Removing winapi-build v0.1.1
      Adding winapi-i686-pc-windows-gnu v0.4.0 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
    Removing winapi-i686-pc-windows-gnu v0.4.0
      Adding winapi-x86_64-pc-windows-gnu v0.4.0 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
    Removing winapi-x86_64-pc-windows-gnu v0.4.0
    Removing ws2_32-sys v0.2.1



$ cargo update -p mio-uds-windows

    Updating git repository `https://github.com/philipktlin/mio-uds-windows.git`
    Updating crates.io index
    Updating git repository `https://github.com/Azure/mio-uds-windows.git`


$ cargo build --all

...
```

At this point, there is only one winapi and only one mio-uds-windows in Cargo.lock:

```

[[package]]
name = "mio-uds"
version = "0.6.7"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "iovec 0.1.2 (git+https://github.com/philipktlin/iovec?branch=arm)",
 "libc 0.2.42 (registry+https://github.com/rust-lang/crates.io-index)",
 "mio 0.6.14 (git+https://github.com/philipktlin/mio?branch=arm)",
]

[[package]]
name = "mio-uds-windows"
version = "0.1.0"
source = "git+https://github.com/philipktlin/mio-uds-windows.git?branch=arm#581fcd3d659ecfe7bf56f4e75e3402dca470c631"
dependencies = [
 "iovec 0.1.2 (git+https://github.com/philipktlin/iovec?branch=arm)",
 "lazycell 1.2.0 (registry+https://github.com/rust-lang/crates.io-index)",
 "log 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "mio 0.6.14 (git+https://github.com/philipktlin/mio?branch=arm)",
 "miow 0.3.1 (git+https://github.com/philipktlin/miow?branch=arm)",
 "winapi 0.3.5 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

[[package]]
name = "miow"
version = "0.3.1"
source = "git+https://github.com/philipktlin/miow?branch=arm#8f63c73f8983df08b66e2aa6d44a9790e91c94bb"
dependencies = [
 "socket2 0.3.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "winapi 0.3.5 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

...

[[package]]
name = "win-logger"
version = "0.1.0"
dependencies = [
 "edgelet-utils 0.1.0",
 "failure 0.1.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "log 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "winapi 0.3.5 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

[[package]]
name = "winapi"
version = "0.3.5"
source = "git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b56cffc81bd94eea7d6a8d006c9dec2f8"
dependencies = [
 "winapi-i686-pc-windows-gnu 0.4.0 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
 "winapi-x86_64-pc-windows-gnu 0.4.0 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b56cffc81bd94eea7d6a8d006c9dec2f8"
```

Reply:
OK, now I can repro what you suggested locally, pushed a fix and started a new build